File: src/mbs/application/comment/rate.py
from typing import Protocol

from mbs.application.common.comment_gateway import CommentReader, CommentRater
from mbs.application.common.exceptions import NotFound, Unauthorized, Forbidden
from mbs.application.common.id_provider import IdProvider
from mbs.application.common.interactor import Interactor
from mbs.domain.models import CommentId
from mbs.domain.services.access import AccessService
import mbs.domain.exceptions as domain_exceptions


class CommentGateway(CommentReader, CommentRater, Protocol):
    pass


class RateComment(Interactor[CommentId, None]):

    def __init__(
            self,
            comment_gateway: CommentGateway,
            access_service: AccessService,
            id_provider: IdProvider
    ):
        self._comment_gateway = comment_gateway
        self._access_service = access_service
        self._id_provider = id_provider

    async def __call__(self, data: CommentId) -> None:
        comment = await self._comment_gateway.get_comment(data)
        if not comment:
            raise NotFound("Комментарий не найден")

        try:
            self._access_service.ensure_can_rate_comment(
                is_auth=self._id_provider.is_auth(),
                permissions=self._id_provider.permissions(),
                user_state=self._id_provider.user_state(),
                comment_state=comment.state
            )
        except domain_exceptions.AccessDenied as error:
            raise Forbidden(str(error))
        except domain_exceptions.AuthenticationError as error:
            raise Unauthorized(str(error))

        await self._comment_gateway.rate_comment(data, self._id_provider.user_id())

File: src/application/comment/rate.rs
use crate::application::common::comment_gateway::CommentGateway;
use crate::application::common::{
    article_gateway::ArticleReader,
    error::AppError,
    id_provider::IdProvider,
    interactor::Interactor
};
use crate::domain::models::rate_state::RateState;
use crate::domain::{
    models::comment::CommentId,
    services::access::ensure_can_rate_comment
};

pub struct RateCommentInput {
    pub id: CommentId,
    pub state: RateState
}

pub struct RateComment<'interactor> {
    pub id_provider: Box<dyn IdProvider>,
    pub comment_gateway: &'interactor dyn CommentGateway,
    pub article_reader: &'interactor dyn ArticleReader
}

impl Interactor<RateCommentInput, ()> for RateComment<'_> {
    async fn execute(&self, input: RateCommentInput) -> Result<(), AppError> {
        let comment = match self.comment_gateway.get_comment(&input.id).await? {
            Some(comment) => comment,
            None => return Err(AppError::NotFound("id".into()))
        };
        
        let article_state = match self.article_reader.get_article_state(&comment.article_id).await? {
            Some(state) => state,
            None => return Err(AppError::Critical("RateComment comment found but article not found".into()))
        };

        ensure_can_rate_comment(
            self.id_provider.permissions(),
            self.id_provider.user_state(),
            &comment.state,
            &article_state
        )?;

        self.comment_gateway.rate_comment(&comment.id, self.id_provider.user_id(), &input.state).await?;
        Ok(())
    }
}
